### ISSUE TITLE
Method getLastInsertId is not covering all possible cases

### ISSUE THREAD

**Issue by spesicbc** [forceedge01/behat-sql-extension#108]
In case that primary key is created by app layer and not named `id`, method in
`https://github.com/forceedge01/behat-sql-extension/blob/master/src/Context/SQLHandler.php#L347`
getLastInsertId defined in
`https://github.com/forceedge01/behat-sql-extension/blob/master/src/Context/DBManager.php#L262`
will trigger an error.
Example:
column name for primary key could be anything else than 'id', like 'guid' etc.
Then, on Postgres, sequence will be tableName_guid_seq, if sequence is defined on database side.
If PKs are generated on app side, then there is no need for those calls at all.

It would work much better if:
1. PK column name will not be hardcoded as id in https://github.com/forceedge01/behat-sql-extension/blob/master/src/Context/DBManager.php#L262 but provided by something like $this->getEntity()->getPrimaryKey()
2. Swap places of https://github.com/forceedge01/behat-sql-extension/blob/master/src/Context/SQLHandler.php#L347 and block under https://github.com/forceedge01/behat-sql-extension/blob/master/src/Context/SQLHandler.php#L352. That way, we can try to fetch inserted id first if it is provided in query, and if not, try to get it from database. Nothing will be broken, it will just give higher prio to PK provided in query, if any.

**Comment by forceedge01** [comment]
@spesicbc thanks very much for reporting. I see what you mean by the id being hardcoded. Given we already dtermine the primary key for the table we can certainly resolve this issue.

For the case of app generated primary keys will have to investigate that avenue a bit farther. You've surely given me enough food for thought though thanks muchly. Will resolve soon but feel free to provide a PR if ur available.

**Comment by forceedge01** [comment]
1) Test with table with primary key other than id.
2) Test with table without primary key.

**Comment by forceedge01** [comment]
@spesicbc I ran some tests as outlined above, I could not replicate the issue you're facing. It may be down to specific versions of software we're using. For your information:

Test ran with:
php v5.6
PostgreSQL 12.2 (Debian 12.2-2.pgdg100+1)

```
CREATE TABLE public.testing (
	id int4 NOT NULL,
	CONSTRAINT testing_pkey PRIMARY KEY (id)
);

create table public.noprimarykey(
	name text
);

create table public.primarykeynotid(
	title char(15),
	name char(25) primary key,
	age int
);
```

```
Feature:
    In order to test out the sql extension
    As the maintainer of the extension
    I want to run a test using the extension

  @postgres
  Scenario:                                          # features/test.feature:7
    And I do not have a "testing" where:             # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | id |
      | 5  |
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: testing
      │ DEBUG >>> SET ENTITY: testing
      │ DEBUG >>> PRIMARY KEY: id
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM testing WHERE "id" = 5
      │ 
    Given I have a "testing" where:                  # Genesis\SQLExtension\Context\SQLContext::iHaveWhere()
      | id |
      | 5  |
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- INSERT -------
      │ DEBUG >>> ENTITY: testing
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Command type set to: insert
      │ DEBUG >>> Executing SQL: INSERT INTO testing ("id") VALUES (5)
      │ DEBUG >>> Last Id: 5
      │ DEBUG >>> Last ID fetched: 5
      │ DEBUG >>> Saving keyword "testing.id" with value "5"
      │ DEBUG >>> Retrieving keyword "testing.id"
      │ DEBUG >>> Retrieved keyword "testing.id" with value "5"
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Executing SQL: SELECT * FROM testing WHERE id = 5
      │ DEBUG >>> Saving keyword "testing.id" with value "5"
      │ 
    And I should have a "testing" with:              # Genesis\SQLExtension\Context\SQLContext::iShouldHaveAWithTable()
      | column | value |
      | id     | 5     |                                                                                                                          
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- EXISTS -------
      │ DEBUG >>> ENTITY: testing
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM testing WHERE "id" = 5
      │ 
    And I do not have a "testing" where:             # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | id |
      | 5  |                                                                                                                                      
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: testing
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM testing WHERE "id" = 5
      │ 
    Then I should not have a "testing" with:         # Genesis\SQLExtension\Context\SQLContext::iShouldNotHaveAWithTable()
      | column | value |
      | id     | 5     |                                                                                                                          
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- NOT-EXISTS -------
      │ DEBUG >>> ENTITY: testing
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "id:5"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM testing WHERE "id" = 5
      │ 
    And I do not have a "noprimarykey" where:        # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | name |
      | aq   |                                                                                                                                    
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: noprimarykey
      │ DEBUG >>> SET ENTITY: noprimarykey
      │ DEBUG >>> PRIMARY KEY: id
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM noprimarykey WHERE "name" = 'aq'
      │ 
    Given I have a "noprimarykey" where:             # Genesis\SQLExtension\Context\SQLContext::iHaveWhere()
      | name |
      | aq   |                                                                                                                                    
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- INSERT -------
      │ DEBUG >>> ENTITY: noprimarykey
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Command type set to: insert
      │ DEBUG >>> Executing SQL: INSERT INTO noprimarykey ("name") VALUES ('aq')
      │ DEBUG >>> Retrieving keyword "noprimarykey.id"
      │ DEBUG >>> -- NO LAST ID RETURNED, NOT CREATING ANY KEYS IN KEYSTORE. --
      │ 
    And I should have a "noprimarykey" with:         # Genesis\SQLExtension\Context\SQLContext::iShouldHaveAWithTable()
      | column | value |
      | name   | aq    |                                                                                                                          
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- EXISTS -------
      │ DEBUG >>> ENTITY: noprimarykey
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM noprimarykey WHERE "name" = 'aq'
      │ 
    And I do not have a "noprimarykey" where:        # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | name |
      | aq   |                                                                                                                                    
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: noprimarykey
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM noprimarykey WHERE "name" = 'aq'
      │ 
    Then I should not have a "noprimarykey" with:    # Genesis\SQLExtension\Context\SQLContext::iShouldNotHaveAWithTable()
      | column | value |
      | name   | aq    |                                                                                                                          
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- NOT-EXISTS -------
      │ DEBUG >>> ENTITY: noprimarykey
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "name:aq"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM noprimarykey WHERE "name" = 'aq'
      │ 
    And I do not have a "primarykeynotid" where:     # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | name   |
      | unique |                                                                                                                                  
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: primarykeynotid
      │ DEBUG >>> SET ENTITY: primarykeynotid
      │ DEBUG >>> PRIMARY KEY: name
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM primarykeynotid WHERE "name" = 'unique'
      │ 
    Given I have a "primarykeynotid" where:          # Genesis\SQLExtension\Context\SQLContext::iHaveWhere()
      | name   |
      | unique |                                                                                                                                  
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- INSERT -------
      │ DEBUG >>> ENTITY: primarykeynotid
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Command type set to: insert
      │ DEBUG >>> Executing SQL: INSERT INTO primarykeynotid ("name") VALUES ('unique')
      │ DEBUG >>> Last Id: unique
      │ DEBUG >>> Last ID fetched: 0
      │ DEBUG >>> Saving keyword "primarykeynotid.name" with value "unique"
      │ DEBUG >>> Retrieving keyword "primarykeynotid.name"
      │ DEBUG >>> Retrieved keyword "primarykeynotid.name" with value "unique"
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Executing SQL: SELECT * FROM primarykeynotid WHERE name = 'unique'
      │ DEBUG >>> Saving keyword "primarykeynotid.title" with value ""
      │ DEBUG >>> Saving keyword "primarykeynotid.name" with value "unique                   "
      │ DEBUG >>> Saving keyword "primarykeynotid.age" with value ""
      │ 
    And I should have a "primarykeynotid" with:      # Genesis\SQLExtension\Context\SQLContext::iShouldHaveAWithTable()
      | column | value  |
      | name   | unique |                                                                                                                         
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- EXISTS -------
      │ DEBUG >>> ENTITY: primarykeynotid
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM primarykeynotid WHERE "name" = 'unique'
      │ 
    And I do not have a "primarykeynotid" where:     # Genesis\SQLExtension\Context\SQLContext::iDontHaveWhere()
      | name   |
      | unique |
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- DELETE -------
      │ DEBUG >>> ENTITY: primarykeynotid
      │ DEBUG >>> Command type set to: delete
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: DELETE FROM primarykeynotid WHERE "name" = 'unique'
      │ 
    Then I should not have a "primarykeynotid" with: # Genesis\SQLExtension\Context\SQLContext::iShouldNotHaveAWithTable()
      | column | value  |
      | name   | unique |
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> ------- NOT-EXISTS -------
      │ DEBUG >>> ENTITY: primarykeynotid
      │ DEBUG >>> Command type set to: select
      │ DEBUG >>> Find external refs in: "name:unique"
      │ DEBUG >>> No external refs found
      │ DEBUG >>> Executing SQL: SELECT * FROM primarykeynotid WHERE "name" = 'unique'
      │ 

1 scenario (1 passed)
15 steps (15 passed)
0m0.18s (13.76Mb)
```

Please provide me with steps to replicate and version of postgres/php you're using. Thanks

**Comment by spesicbc** [comment]
Thanks for your quick reponse. Ok, here we go:

Ubuntu: 16.04.6 LTS
Postgres: postgres (PostgreSQL) 9.6.17

CREATE TABLE public.testtable
(
    guid uuid NOT NULL,
    name character varying COLLATE pg_catalog."default",
    CONSTRAINT testtable_pkey PRIMARY KEY (guid)
)

***
Scenario: Insert new record.
    Given I have a "testtable" where:
    | guid | name |
    | 00000000-0000-0000-4444-00000000000b | New name |

Scenario: Insert new record.        # tests/Acceptance/features/Regression/test.feature:107
      Given I have a "testtable" where: # tests/Acceptance/features/Regression/test.feature:108
        │ DEBUG >>> Find external refs in: "guid:00000000-0000-0000-4444-00000000000b,name:New name"
        │ DEBUG >>> No external refs found
        │ DEBUG >>> ------- INSERT -------
        │ DEBUG >>> ENTITY: testtable
        │ DEBUG >>> SET ENTITY: testtable
        │ DEBUG >>> PRIMARY KEY: guid
        │ DEBUG >>> Find external refs in: "guid:00000000-0000-0000-4444-00000000000b,name:New name"
        │ DEBUG >>> No external refs found
        │ DEBUG >>> Command type set to: insert
        │ DEBUG >>> Executing SQL: INSERT INTO testtable ("guid", "name") VALUES ('00000000-0000-0000-4444-00000000000b', 'New name')
        │ 
        Unable to insert data into table 'testtable', Error Warning: PDO::lastInsertId(): SQLSTATE[42P01]: Undefined table: 7 ERROR:  relation "testtable_id_seq" does not exist in vendor/genesis/behat-sql-extension/src/Context/DBManager.php line 262, Entity: Genesis\SQLExtension\Context\Representations\Entity Object
        (
            [rawInput:Genesis\SQLExtension\Context\Representations\Entity:private] => testtable
            [tableName:Genesis\SQLExtension\Context\Representations\Entity:private] => testtable
            [databaseName:Genesis\SQLExtension\Context\Representations\Entity:private] => 
            [schemaName:Genesis\SQLExtension\Context\Representations\Entity:private] => 
            [primaryKey:Genesis\SQLExtension\Context\Representations\Entity:private] => guid
            [requiredColumns:Genesis\SQLExtension\Context\Representations\Entity:private] => Array
                (
                )
        
        )
         (Genesis\SQLExtension\Context\Exceptions\InsertException)

**Comment by forceedge01** [comment]
Cool I'll test using the above versions. Can u confirm the php version for completeness sake please?

**Comment by spesicbc** [comment]
PHP 7.2.30-1+ubuntu16.04

**Comment by forceedge01** [comment]
@spesicbc preliminary fix to this issue has been raised here https://github.com/forceedge01/behat-sql-extension/issues/108, I need to test this a bit further in order to release it but you can pull the branch and see if this resolves all your issues. All scenarios have been tested with postgres. FYI  - The issue was surprisingly stemming from php7.2's pdo pgsql library. On php5.6 this doesn't happen.

**Comment by spesicbc** [comment]
Thank you @forceedge01. I've tested insert/update/delete and it seams that everything is fixed now. I will inform you if I encounter any other issues.
From my point of view, you can merge it to master.

**Comment by forceedge01** [comment]
Thanks for reporting @spesicbc 

Fix released in v 8.5.1
